log-collector/collector.go

```diff
@@ -0,0 +1,198 @@
+package collector
+
+import (
+	"io"
+	"log"
+	"path/filepath"
+	"sync"
+
+	"github.com/coreos/ktestutil/log-collector/pkg/fluentd"
+	"github.com/coreos/ktestutil/log-collector/pkg/local"
+	"github.com/coreos/ktestutil/log-collector/pkg/s3"
+	"github.com/coreos/ktestutil/log-collector/pkg/scp"
+
+	"k8s.io/client-go/kubernetes"
+)
+
+// Output interface describes where the log file is uploaded
+//
+// Put accepts an 'io.ReadSeeker' of the log file and the 'filename' at the destination.
+// Put returns the 'absolute (accessible) location' of the log file at the destination
+// and an error if the file put was unsuccessful
+type Output interface {
+	Put(io.ReadSeeker, string) (string, error)
+}
+
+type Collector struct {
+	scpConfig *scp.Config
+	namespace string
+	k8s       kubernetes.Interface
+
+	Output Output
+}
+
+// Config defines configuration options for the Collector
+//
+// Requires K8sClient and Namespce.
+// If RemoteKeyFile is empty uses SSH_AUTH_SOCK to establish ssh connection.
+// If RemoteUser is empty 'core' is used as default user for ssh connection.
+// If RemotePort is empty '22' is used as default port for ssh connection.
+type Config struct {
+	RemoteUser    string
+	RemotePort    int32
+	RemoteKeyFile string
+	K8sClient     kubernetes.Interface
+	Namespace     string
+}
+
+// New returns *Collector given a *Config
+func New(c *Config) *Collector {
+	return &Collector{
+		scpConfig: &scp.Config{
+			User:            c.RemoteUser,
+			Port:            c.RemotePort,
+			IdentifyKeyFile: c.RemoteKeyFile,
+		},
+		namespace: c.Namespace,
+		k8s:       c.K8sClient,
+	}
+}
+
+// OutputToLocal sets the Collector output to local dir dstDir
+func (cr *Collector) OutputToLocal(dstDir string) error {
+	l, err := local.New(&local.Config{
+		Dir: dstDir,
+	})
+	if err != nil {
+		return err
+	}
+
+	cr.Output = l
+	return nil
+}
+
+// OutputToS3 sets the Collector output to S3 bucket, given proper credentials
+func (cr *Collector) OutputToS3(keyId, keySecret, region, bucketName, bucketPrefix string) error {
+	s3, err := s3.New(&s3.Config{
+		AccessKeyId:     keyId,
+		AccessKeySecret: keySecret,
+		BucketName:      bucketName,
+		BucketPrefix:    bucketPrefix,
+		Region:          region,
+	})
+	if err != nil {
+		return err
+	}
+
+	cr.Output = s3
+	return nil
+}
+
+// Start creates the fluentd master and worker assets for log collection.
+// it also registers the node that is running the fluentd master
+func (cr *Collector) Start() error {
+	if err := fluentd.CreateAssets(cr.k8s, cr.namespace); err != nil {
+		return err
+	}
+
+	host, err := fluentd.GetNodeAddressWithMaster(cr.k8s, cr.namespace)
+	if err != nil {
+		return err
+	}
+	cr.scpConfig.Host = host
+	return nil
+}
+
+// CollectPodLogs fetches the log file(s) for the pods matching basic shell file name pattern
```

Comment:
basic shell file name patter == glob? if yes, then change pod string to pobGlob string?